Description: tslint is not yet packaged
Author: Laszlo Boszormenyi (GCS) <gcs@debian.org>
Last-Update: 2016-01-07

---

--- typescript-1.7.5.orig/Jakefile.js
+++ typescript-1.7.5/Jakefile.js
@@ -4,7 +4,7 @@ var fs = require("fs");
 var os = require("os");
 var path = require("path");
 var child_process = require("child_process");
-var Linter = require("tslint");
+//var Linter = require("tslint");
 
 // Variables
 var compilerDirectory = "src/compiler/";
