--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = $(shell awk '/Version:/ { print $$2 }' logrotate.spec)
 OS_NAME = $(shell uname -s)
 LFS = $(shell echo `getconf LFS_CFLAGS 2>/dev/null`)
-CFLAGS = -Wall -D_GNU_SOURCE -D$(OS_NAME) -DVERSION=\"$(VERSION)\" $(RPM_OPT_FLAGS) $(LFS)
+CFLAGS = -Wall -D_GNU_SOURCE -D$(OS_NAME) -DVERSION=\"$(VERSION)\" $(RPM_OPT_FLAGS) $(LFS) -DCOMPRESS_COMMAND=\"/usr/bin/gzip\" -DUNCOMPRESS_COMMAND=\"/usr/bin/gunzip\"
 PROG = logrotate
 MAN = logrotate.8
 MAN5 = logrotate.conf.5
