--- imspector/Makefile	2012-02-17 13:46:22.000000000 +0000
+++ imspector.new/Makefile	2012-02-17 13:39:43.000000000 +0000
@@ -7,7 +7,7 @@
 
 ##########
 # Location of openssl installation
-SSL_DIR = /usr
+SSL_DIR = /usr/ssl
 
 ##########
 # If using FreeBSD or OpenBSD COMMENT the below lines
@@ -171,8 +171,8 @@
 	-install imspector.conf $(DESTDIR)/$(PREFIX)/etc/imspector/imspector.conf
 	-install badwords.txt $(DESTDIR)/$(PREFIX)/etc/imspector/badwords.txt
 	-install acl.txt $(DESTDIR)/$(PREFIX)/etc/imspector/acl.txt
-	-mkdir -p /var/log/imspector
-	-mkdir -p /var/lib/imspector
+#	-mkdir -p /var/log/imspector
+#	-mkdir -p /var/lib/imspector
 
 install-cert:
 	-mkdir -p $(DESTDIR)/$(PREFIX)/etc/imspector
