Author: Sergey Poznyakoff <gray@gnu.org.ua>
Description: Fix conflicting declarations.
Forwarded: not-needed

diff --git a/include/mailutils/cpp/list.h b/include/mailutils/cpp/list.h
index 8bb6adc..45f38dd 100644
--- a/include/mailutils/cpp/list.h
+++ b/include/mailutils/cpp/list.h
@@ -25,9 +25,6 @@
 #include <mailutils/cpp/error.h>
 #include <mailutils/cpp/iterator.h>

-typedef int mu_list_action_t (void*, void*);
-typedef int (*mu_list_comparator_t) (const void*, const void*);
-
 namespace mailutils
 {

@@ -64,7 +61,7 @@ class List
   void to_array (void** array, size_t count, size_t* pcount);
   void locate (void* item, void** ret_item);

-  void apply (mu_list_action_t* action, void* cbdata);
+  void apply (mu_list_action_t action, void* cbdata);
   mu_list_comparator_t set_comparator (mu_list_comparator_t comp);
   mu_list_destroy_item_t set_destroy_item (mu_list_destroy_item_t mu_destroy_item);

diff --git a/libmu_cpp/list.cc b/libmu_cpp/list.cc
index d341a1c..0e4c0e0 100644
--- a/libmu_cpp/list.cc
+++ b/libmu_cpp/list.cc
@@ -199,7 +199,7 @@ List :: size ()
 }

 void
-List :: apply (mu_list_action_t* action, void* cbdata)
+List :: apply (mu_list_action_t action, void* cbdata)
 {
   int status = mu_list_foreach (mu_list, action, cbdata);
   if (status)
