--- a/driver/bypass_control.c
+++ b/driver/bypass_control.c
@@ -69,7 +69,6 @@
 static int pca9555_probe(struct i2c_client *client,
                         const struct i2c_device_id *id)
 {
-        client->driver = &ca_bypass_driver;
         number_of_bypass_device++;
 #ifdef CONFIG_CA_BOARD_DEBUG
         bypass_setup_fs(client, DEFATTR);
@@ -130,7 +129,6 @@
 static int pca9555_probe(struct i2c_client *client,
                         const struct i2c_device_id *id)
 {
-	client->driver = &ca_bypass_driver;
 	number_of_bypass_device++;
 #ifdef CONFIG_CA_BOARD_DEBUG    
 	bypass_setup_fs(client, DEFATTR);
@@ -219,7 +217,6 @@
                 memcpy(&eeprom_client, new_client, sizeof(struct i2c_client));
                 return 0;
         } else {
-                new_client->driver = &ca_bypass_driver;
                 new_client->flags = 0;
 
                 if (kind < 0) {
@@ -344,7 +341,6 @@
 		memcpy(&eeprom_client, new_client, sizeof(struct i2c_client));
 		return 0;
 	} else {
-		new_client->driver = &ca_bypass_driver;
 		new_client->flags = 0;
 
 		/* Fill in the remaining client fields and put it into the global list */
