Description: Use correct path to VERSION
Author: Jonas Smedegaard <dr@jones.dk>
Last-Update: 2014-07-07

--- a/lib/sass/root.rb
+++ b/lib/sass/root.rb
@@ -3,5 +3,5 @@
   # This may be overridden by the package manager
   # if the lib directory is separated from the main source tree.
   # @api public
-  ROOT_DIR = File.expand_path(File.join(__FILE__, "../../.."))
+  ROOT_DIR = File.expand_path(File.join(__FILE__, ".."))
 end
